### PR TITLE
Change master->primary/leader/member

### PR DESCRIPTION
**Pull request by CyberDem0n**
keep as much backward compatibility as possible.

Following changes were made:
1. All internal checks are performed as `role in ('master', 'primary')`
2. All internal variables/functions/methods are renamed
3. `GET /metrics` endpoint returns `patroni_primary` in addition to `patroni_master`.
4. Logs are changed to use leader/primary/member/remote depending on the context
5. Unit-tests are using only role = 'primary' instead of 'master' to verify that 1 works.
6. patronictl still supports old syntax, but also accepts `--leader` and `--primary`.
7. `master_(start|stop)_timeout` is automatically translated to `primary_(start|stop)_timeout` if the last one is not set.
8. updated the documentation and some examples

Future plan: in the next major release switch role name from `master` to `primary` and maybe drop `master` altogether.
The Kubernetes implementation will require more work and keep two labels in parallel. Label values should probably be configurable as described in https://github.com/zalando/patroni/issues/2495.

### CODE REVIEW

**Comment by CyberDem0n** [comment]
@matthbakeredb, I noticed you liked the PR. Do you want to take a look/try/test it? :)

**Comment by matthbakeredb** [comment]
> @matthbakeredb, I noticed you liked the PR. Do you want to take a look/try/test it? :)

hehe yes of course :smile:

File: tests/test_api.py
Comment (matthbakeredb) [inline tests/test_api.py:215]: Perhaps should check the content of this query as it's returning `master`

```
postgres@patroni2:~$ curl -s http://127.0.0.1:8008/primary|jq .role
"master"
```

File: patroni/ctl.py
Comment (matthbakeredb) [inline patroni/ctl.py:250]: Looks like `master` is still the actual value, so still a "first class citizen", rather than replaced with `primary`. Shouldn't this be replaced both here and where it is being retrieved from?

**Review by matthbakeredb — COMMENTED** [review]
Some thoughts; So the config and ctl cli is backwards compatible afaict but you're removing the /master api endpoint? Is this intended to be transitional, ie master to be removed entirely in a future release? If this is to be released using a new major version (non-backwards compatible) then perhaps `master` could be removed entirely. If it is transitional then I'd expect all interfaces to support the new and old naming, with the new naming as a "first class citizen" in all defaults, internal refs, docs, etc.

Do we want to leave it to the package maintainers to include migration tooling for the required changes to config?

Otherwise, everything looks good. Runs OK locally, behave tests passing, basic interaction seems good.

File: patroni/ctl.py
Comment (CyberDem0n) [reply patroni/ctl.py:250]: Yes, in this place we can reverse primary<->master in dict. 
But, REST API will still return role=master and so will be stored in the member key in DCS.
We can't simply change everything in one shot because it will break backward compatibility.
For example, during leader race all versions up to 2.1.5 are checking for role==master: https://github.com/zalando/patroni/blob/v2.1.5/patroni/ha.py#L42

File: tests/test_api.py
Comment (CyberDem0n) [reply tests/test_api.py:215]: Yes, it is expected that new versions will still use "master" for a while.
Right now we change only tests in order to verify that the actual code correctly handles both cases, when role == 'master' or role == 'primary'.
Now we are planning a release 3.0.
My plan is to completely get rid from master only in 3.1, where in release notes we could write - "you should be running at least 3.0.0" in order to safely upgrade to "3.1.0" (mainly because of https://github.com/zalando/patroni/blob/v2.1.5/patroni/ha.py#L42). Upgrading directly from older version is sort of also possible, but there will be very easy to screw things up :(

Comment (matthbakeredb) [reply tests/test_api.py:215]: I understand that it is a big change, bumping the major version notifies the user of backwards incompatible changes. Perhaps, if this change is backwards compatible this could go into 2.2, by allowing the user to get used to using the new syntax/interfaces whilst fully supporting all the old, with the main switch to not having `master` in 3.0? That would be more in-line with semver imo.

File: patroni/ctl.py
Comment (matthbakeredb) [reply patroni/ctl.py:250]: Perhaps there could be data migration in the dcs code that renames stored values? That could be optional in backwards compatible versions and enforced in non-backwards compat versions.

File: tests/test_api.py
Comment (CyberDem0n) [reply tests/test_api.py:215]: This is not only about breaking the API, which indeed we don't break it yet.
There are two huge features recently merged: DCS failsafe and Citus support.
IMO, it deserves 3.0 :)

File: patroni/ctl.py
Comment (CyberDem0n) [reply patroni/ctl.py:250]: There is no data migration required.
Every node maintains its own member key.
The only problem is that the client tooling and new Patroni are required to understand new values.

File: tests/test_api.py
Comment (matthbakeredb) [reply tests/test_api.py:215]: ah ha! Fair enough.

**Review by matthbakeredb — APPROVED** [review]
:+1:

**Comment by CyberDem0n** [comment]
> So the config and ctl cli is backwards compatible afaict but you're removing the /master api endpoint?

No, it is not removed. All kind of "monitoring" endpoints are handled in the "do_GET()" method. It supports /primary since a long time.

> Is this intended to be transitional, ie master to be removed entirely in a future release?

Yes, in some future release it will be removed entirely.

> If this is to be released using a new major version (non-backwards compatible) then perhaps master could be removed entirely. If it is transitional then I'd expect all interfaces to support the new and old naming, with the new naming as a "first class citizen" in all defaults, internal refs, docs, etc.

That's more or less what I did. Docs are using only the "primary" or "leader".
There are only three places left where the "master" is used:
- REST API response: `{"role":"master", ...}`
- "/members/$name" in DCS: `{"role":"master", ...}`
- K8s "role" label on the leader pod

When everything is migrated to 3.0+, we can change REST API and DCS to `{"role":"primary", ...}`, entirely remove "master" from the code, and release 3.1.

With K8s it is way more complex. More details is in the #2495 

> Do we want to leave it to the package maintainers to include migration tooling for the required changes to config?

There is nothing package maintainers could do about it. Also, many people install Patroni with pip.

> Otherwise, everything looks good. Runs OK locally, behave tests passing, basic interaction seems good.

Well, yeah with behave tests are out of question, because we run them for every commit :)
Thank you for you

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**
